### PR TITLE
[JEP-230] Remove SSH Server Plugin from core

### PR DESCRIPTION
**Pull request by daniel-beck**
Followup to #5049 and #5327: As described in https://github.com/jenkinsci/jenkins/pull/5049#issuecomment-816489800, there is IMO no real reason to consider this a "detached plugin":

* It's probably not a commonly used feature, and can easily be installed again without data loss (separate descriptor config file)
* It doesn't make sense as an implied dependency, because existing plugins would not depend on it implicitly with old core baselines, so this just adds an implied dependency for no real reason.

This change makes this removal more in line with https://plugins.jenkins.io/cctray-xml/ which was simply removed from core and replaced with a compatible plugin.

There needs to be more due diligence here (for the second list item especially, unsure how dependencies on modules work), but filing this for discussion. WDYT?

Untested.

### Proposed changelog entries

* No longer bundle the SSH Server Plugin. Users of Jenkins 2.281 and earlier who update Jenkins and currently use the built-in SSH Server functionality to run Jenkins CLI commands can install the plugin again via the plugin manager.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by jvz — APPROVED** [review]
I think this makes sense. The WebSocket option for the CLI works fairly well without additional configuration, so the sshd makes sense to remove from the default install here.

**Review by kuisathaverat — APPROVED** [review]
A lighter Jenkins core

**Review by timja — APPROVED** [review]

**Comment by daniel-beck** [comment]
This is likely a breaking change; plugins do not declare dependencies on modules in the manifest like they do dependencies on plugins. So modules are on the classpath at runtime.

This module is referenced in about a dozen or so pom.xml files with provided scope (rather than test), which is probably reasonable to clean up?

**Comment by oleg-nenashev** [comment]
Let's put it on hold until @jglick can review it as the JEP-230 author

**Comment by jglick** [comment]
> It doesn't make sense as an implied dependency, because existing plugins would not depend on it implicitly with old core baselines

As you found, `git-server` and a few miscellaneous/abandoned plugins do rely on its presence, without an explicit mention in `Plugin-Dependencies`, only implied by `Jenkins-Version`—just like any other detached plugin. The only difference I can see is that it is easier to find these cases by grepping `pom.xml` for `provided` deps.

Now if you want to more aggressively clean up entries in `war/pom.xml` (_not_ `split-plugins.txt`) because they have been detached long ago, as you have been doing, that is fine, and if you want to clean up recently detached plugins that is OK so long as you track down all the plugins on the update center known to be using the dep and update them to a new core baseline + explicit (`compile`-scope) `sshd` dep.

**Review by jglick — CHANGES_REQUESTED** [review]
Incompatible, requires some plugin changes to be prepared and released.

**Comment by jglick** [comment]
> probably not a commonly used feature

Perhaps you were referring here to SSH mode of the Jenkins CLI? Probably not too common, and I agree that users interested in this can install the plugin explicitly.

`git-server` is a dep of `workflow-cps-global-lib`, used in the original mode of the plugin whereby you would push library changes to `workflowLibs.git` inside Jenkins. This has of course been long deprecated in favor of external libraries, but who knows how many people are still relying on it. See: https://github.com/jenkinsci/jep/blob/master/jep/230/README.adoc#cleanup-options

**Comment by daniel-beck** [comment]
Not a good idea after all 😭

**Comment by jglick** [comment]
To be clear, I think this would be great _if_ all the maintained plugins depending on the SSH server can get their POMs refreshed first. Then we would simply mention in release notes for the next core version that you are expected to apply some plugin updates first, or risk some linkage errors.

**Comment by daniel-beck** [comment]
@jglick Agree, but needs more time than I have for this right now.

**Comment by kuisathaverat** [comment]
@daniel-beck @jglick I've made those PRs for the plugins that are alive https://github.com/pulls?q=is%3Aopen+is%3Apr+author%3Akuisathaverat+archived%3Afalse+jenkins-64107+ I need help from the maintainers in a couple.

Also, https://github.com/jenkinsci/sshd-plugin/pull/37 is important there is a regression related to the bump of the Apache Mina SSHD on the jenkins CLI [JENKINS-65273](https://issues.jenkins.io/browse/JENKINS-65273)

**Comment by timja** [comment]
I've pinged the maintainers

**Pull request closed without merging**
